### PR TITLE
[Issue-3617] Enables FlyteFiles, FlyteDirectors, and StructuredDatasets inputs in papermill plugin

### PR DESCRIPTION
**Pull request by peridotml**
# TL;DR
This is a first draft for allowing the papermill plugin to access more complex inputs. 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Right now the papermill plugin can't use `FlyteFile` or `StructuredDataset` because of a Papermill limitation that everything is json serializable.

My strategy is mostly copying `record_outputs`. These are the steps

1.  Reserializing the inputs and save to local disk
2. Add a helper function `read_input` where users must give the input type (maybe there is a smart fix)
3. `read_input` loads the protos and uses the type to convert it to the correct format

Basically I am using flyte to serialize the inputs locally (see comments), then you just need to do this. 

```python
### Params
pi = (filled in by papermill)

### In a new cell
pi = read_input(pi, float)
```
Screenshot 2023-05-02 at 1 42 43 PM

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3617

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 1]
```diff
@@ -256,7 +259,8 @@ def execute(self, **kwargs) -> Any:
         """
         logger.info(f"Hijacking the call for task-type {self.task_type}, to call notebook.")
         # Execute Notebook via Papermill.
-        pm.execute_notebook(self._notebook_path, self.output_notebook_path, parameters=kwargs, log_output=self._stream_logs)  # type: ignore
+        seralized_kwargs = serialize_inputs(**kwargs)
```
Comment (peridotml) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:262]: It would be nice to only serialize the datatypes that need to be serialized. I wasn't sure how to check for that.

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 2]
```diff
@@ -307,3 +311,35 @@ def record_outputs(**kwargs) -> str:
...
+        tmp = tempfile.mktemp(suffix="bin")
+        utils.write_proto_to_file(lit.to_flyte_idl(), tmp)
+        outputs[k] = tmp
+
+    return outputs
+
+
+def read_input(path: str, dtype: T) -> T:
+
+    if type(path) == dtype:
```
Comment (peridotml) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:338]: Sometimes users might fill in default values in the parameters cell in a papermill notebook. We still want the notebooks to work in this case.

**Review by peridotml — COMMENTED** [review]
I had to learn a bunch about the type transformer to do this, so that was fun. There are clearly some rough edges, so I would appreciate suggests.

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py (hunk 1 above)
Comment (pingsutw) [reply plugins/flytekit-papermill/flytekitplugins/papermill/task.py:262]: something like:
```python
for k, v in kwargs.items():
  if is_instance(v, StructuredDataset):
    kwargs[k] = serialize_inputs(v)
```

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 3]
```diff
@@ -255,6 +259,10 @@ def execute(self, **kwargs) -> Any:
         singleton
         """
         logger.info(f"Hijacking the call for task-type {self.task_type}, to call notebook.")
+        for k, v in kwargs.items():
+            if isinstance(v, (FlyteFile, FlyteDirectory, StructuredDataset)):
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:263]: nit: store this in a global variable in the top-level scope of this module e.g.

```
SAVE_TO_LITERAL = (FlyteFile, FlyteDirectory, StructuredDataset)
```

It's effectively the same, but will be easier to extend this to perhaps other types we want to save to literal.

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 4]
```diff
@@ -274,7 +283,7 @@ def execute(self, **kwargs) -> Any:
                 v = TypeEngine.to_python_value(ctx=FlyteContext.current_context(), lv=m[k], expected_python_type=type_v)
                 output_list.append(v)
             else:
-                raise RuntimeError(f"Expected output {k} of type {v} not found in the notebook outputs")
+                raise RuntimeError(f"Expected output {k} of type {type_v} not found in the notebook outputs")
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:286]: raise TypeError here instead of RuntimeError

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 5]
```diff
@@ -307,3 +316,51 @@ def record_outputs(**kwargs) -> str:
         lit = TypeEngine.to_literal(ctx, python_type=type(v), python_val=v, expected=expected)
         m[k] = lit
     return LiteralMap(literals=m).to_flyte_idl()
+
+
+def save_literal_to_file(input: Any) -> str:
+    """
+    Serializes an input
+    """
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:324]:
```suggestion
    """
    Serializes an input
    
    :param input: the python input value to serialize.
    """
```

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 6]
```diff
@@ -307,3 +316,51 @@ def record_outputs(**kwargs) -> str:
...
+    lit = TypeEngine.to_literal(ctx, python_type=type(input), python_val=input, expected=expected)
+
+    tmp_file = tempfile.mktemp(suffix="bin")
+    utils.write_proto_to_file(lit.to_flyte_idl(), tmp_file)
+    return tmp_file
+
+
+def read_input(path: str, dtype: T) -> T:
+    """
+    Reads a Flyte literal from a file
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:336]: add parameter docstrings (see above)

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 7]
```diff
@@ -307,3 +316,51 @@ def record_outputs(**kwargs) -> str:
...
+    proto = utils.load_proto_from_file(_pb2_Literal, path)
+    lit = Literal.from_flyte_idl(proto)
+    ctx = FlyteContext.current_context()
+    python_value = TypeEngine.to_python_value(ctx, lit, dtype)
+    return python_value
+
+
+def read_flytefile(path: str) -> T:
+    """
+    Use this method to read a FlyteFile literal from a file.
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:350]: add parameter docstrings

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 8]
```diff
@@ -307,3 +316,51 @@ def record_outputs(**kwargs) -> str:
...
+    """
+    return read_input(path=path, dtype=FlyteFile)
+
+
+def read_flytedirectory(path: str) -> T:
+    """
+    Use this method to read a FlyteDirectory literal from a file.
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:357]: add parameter docstrings

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 9]
```diff
@@ -307,3 +316,51 @@ def record_outputs(**kwargs) -> str:
...
+    """
+    return read_input(path=path, dtype=FlyteDirectory)
+
+
+def read_structureddataset(path: str) -> T:
+    """
+    Use this method to read a StructuredDataset literal from a file.
```
Comment (cosmicBboy) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:364]: add parameter docstrings

**Review by cosmicBboy — APPROVED** [review]
